packages/server/src/constraints.ts

```diff
@@ -0,0 +1,114 @@
+import { DAY, HOUR, cborEncode, check } from '@atproto/common'
+import * as plc from '@did-plc/lib'
+import { ServerError } from './error'
+import { parseDidKey } from '@atproto/crypto'
+
+export function assertValidIncomingOp(
+  op: unknown,
+): asserts op is plc.OpOrTombstone {
+  const byteLength = cborEncode(op).byteLength
+  if (byteLength > 4000) {
+    throw new ServerError(400, 'Operation too large')
+  }
+  if (!check.is(op, plc.def.opOrTombstone)) {
+    throw new ServerError(400, `Not a valid operation: ${JSON.stringify(op)}`)
+  }
+  if (op.type === 'plc_tombstone') {
+    return
+  }
+  if (op.alsoKnownAs.length > 10) {
+    throw new ServerError(400, 'To many alsoKnownAs items (max 10)')
+  }
+  const akaDupe: Record<string, boolean> = {}
+  for (const aka of op.alsoKnownAs) {
+    if (aka.length > 256) {
+      throw new ServerError(400, `alsoKnownAs field too long (max 256): ${aka}`)
+    }
+    if (akaDupe[aka]) {
+      throw new ServerError(400, `duplicate alsoKnownAs field: ${aka}`)
+    } else {
+      akaDupe[aka] = true
+    }
+  }
+  if (op.rotationKeys.length > 5) {
```

Comment:
I'd probably be a touch more permissive for rotation keys, like 10?

- PDS might have 2x rotation keys for easier ops, one online and one cold/offline (we should probably do this!)
- user has yubikey and re-uses an existing offline paper key as rotation keys
- user delegates another rotation to a friend
- user delegates rotation to a trusted third party service or lawyer or something

That's a bit of a stretch, but not crazy for an activist or politician or newspaper, and gets to 6.